client-java/src/main/java/uk/gov/gchq/palisade/client/internal/resource/WebSocketClient.java

```diff
@@ -0,0 +1,209 @@
+
+        /**
+         * Returns the base web socket uri
+         *
+         * @return the base web socket uri
+         */
+        URI getUri();
+
+        /**
+         * Returns the object mapper used for (de)serialisation of websocket messages
+         *
+         * @return the object mapper used for (de)serialisation of websocket messages
+         */
+        ObjectMapper getObjectMapper();
+
+        /**
+         * Returns the HTTP client that should be used by the {@code WebSocketClient}
+         *
+         * @return the HTTP client that should be used by the {@code WebSocketClient}
+         */
+        HttpClient getHttpClient();
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketClient.class);
+
+    private final ResourceClientSetup setup;
+    private final BlockingQueue<WebSocketMessage> next = new LinkedBlockingQueue<>(1);
+
+    private WebSocket webSocket;
+
+
+    /**
+     * A {@code ResourceClient} manages the passing of messages to/from a websocket
+     * server
+     *
+     * @param setup The setup instance
+     */
+    public WebSocketClient(final ResourceClientSetup setup) {
+        this.setup = Checks.checkNotNull(setup);
+    }
+
+    /**
+     * Helper method to create a {@link Item} using a builder function
+     *
+     * @param func The builder function
+     * @return a newly created {@code RequestId}
+     */
+    public static WebSocketClient createResourceClient(final UnaryOperator<ResourceClientSetup.Builder> func) {
+        return new WebSocketClient(func.apply(new ResourceClientSetup.Builder()).build());
+    }
+
+    /**
+     * Retrieves and removes the next message, waiting up to the specified wait time
+     * if necessary for a message to become available.
+     *
+     * @param timeout how long to wait before giving up, in units of {@code unit}
+     * @param unit    a {@code TimeUnit} determining how to interpret the
+     *                {@code timeout} parameter
+     * @return the the next message, or {@code null} if the specified waiting time
+     *         elapses before a message is available
+     */
+    public WebSocketMessage poll(final long timeout, final TimeUnit unit) {
+        try {
+            return next.poll(timeout, unit);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Thread interrupted while taking next message from queue");
+        }
+    }
+
+    /**
+     * Connect to the server and start communications
+     *
+     * @return this for fluent usage
+     */
+    public WebSocketClient connect() {
+
+        // as soon as a client subscribes, the connection to the filtered resource
+        // service should be instantiated and messages should start to be emitted
+
+        // note that %t has been encoded as %25t in the URI
+
+        var replacedUri = URI.create(getUri().toString().replace("%25t", getToken()));
```

Comment:
Is this simply because `%25` percent-encodes to `%`? If so, either getUri() is a string, so you can use `%t` in both cases, or it's a uri, so in this case the .toString() produces a `%` for substitution? Or does URI .toString do a percent-decode too?
All in all, could be confusing to your end-user

Reply:
I was using strings until feedback suggested to store them as URI's. This is completely hidden from the end user. If a user did want to provide a different path (for testing), they would use %t as it would be encoded internally.